File: robot/src/jade/JadePair.java
package jade;

/**
 *  The JadePair acts like a tuple, it stores both a String value of first and second.
 *  This normally acts like a key pair with the first being the key and the second being the data it pertains to.
 * 
 * @author  Joseph Beck
 * @version 0.1
 * @since   2023
 */
public class JadePair {

    /**
     * String that stores the first variable.
     */
    private String first;
    /**
     * String that stores the second variable.
     */
    private String second;

    /**
     * Takes in two string parameters and initialises the object.
     * 
     * @param String first
     * @param String second
     * @return none
     */
    public JadePair(String first, String second) {
        this.first = first;
        this.second = second;
    }

    /**
     * Takes in no parameters and initialises a null instance of JadePair.
     * 
     * @return none
     */
    public JadePair() {
        this.first = null;
        this.second = null;
    }

    /**
     * Takes in  two string parameters and initialises the object.
     * 
     * @param String first
     * @param String second
     * @return none
     */
    public void replace(JadePair pair) {
        this.first = pair.getFirst();
        this.second = pair.getSecond();
    }

    /**
     * Returns the value of the first variable.
     * 
     * @return String
     */
    public String getFirst() {
        return this.first;
    }

    /**
     * Sets the value of the first variable.
     * 
     * @param String first
     * @return none
     */
    public void setFirst(String first) {
        this.first = first;
    }

    /**
     * Returns the value of the second variable.
     * 
     * @return String
     */
    public String getSecond() {
        return this.second;
    }

    /**
     * Sets the value of the second variable.
     * 
     * @param String second
     * @return none
     */
    public void setSecond(String second) {
        this.second = second;
    }

    /**
     * Returns a string representation of this Jade Pair object.
     * 
     * @return String
     */
    @Override
    public String toString() {
        return first + " : " + second;
    }
}

File: client/jade/jade_pair.rb
# frozen_string_literal: true

# The JadePair is essential to the JadeObject.
# It stores two stings of first and second.
# Commonly the first is the variable name and the second is the variable data.
# JadePairs are used commonly throughout both the JadeSerializer and JadeDeserializer too.
#
# @author Joseph Beck
# @!visibility public
class JadePair

  public

  # First and second are commonly used outside of the JadePair itself and therefore require an attribute accessor.
  #
  # @!attribute :first, :second
  # @!visibility public
  attr_accessor :first, :second

  # Initializes the instance of both first and second.
  # These default to nil if nothing is provided in the parameters.
  #
  # @param [String, String]
  # @return [NilClass]
  # @!visibility public
  def initialize(first = nil, second = nil)
    @first = first
    @second = second
  end

  # Allows both the values of first and second to be replaced.
  #
  # @param [String, String]
  # @return [NilClass]
  # @!visibility public
  def replace(first, second)
    @first = first
    @second = second
    nil
  end

  # Allows the value of first to be replaced.
  #
  # @param [String]
  # @return [NilClass]
  # @!visibility public
  def replace_first(first)
    @first = first
    nil
  end

  # Allows the value of second to be replaced.
  #
  # @param [String]
  # @return [NilClass]
  # @!visibility public
  def replace_second(second)
    @second = second
    nil
  end

  # to_array provides an array of both the values of first and second.
  # This may be useful in some cases of JadeObject construction.
  #
  # @param [NilClass]
  # @return [String[]]
  # @!visibility public
  def to_array
    [@first, @second]
  end

  # to_str converts the current instance of the JadePair to a string.
  # Useful when wanting a simple representation of the current JadePair.
  #
  # @param [NilClass]
  # @return [String]
  # @!visibility public
  def to_str
    "#{@first} : #{@second}"
  end

  # Checks whether the current instance of JadePair is empty
  # Useful for preventing uses of nil JadePairs
  #
  # @param [NilClass]
  # @return [Boolean]
  # @!visibility public
  def empty?
    @first.nil? && @second.nil?
  end

  # Checks whether the current instance of JadePair is equal to another JadePair
  # Useful for checking JadePair equivalence
  #
  # @param [JadePair]
  # @return [Boolean]
  # @!visibility public
  def eql?(other)
    ((@first.eql? other.first) &&
     (@second.eql? other.second))
  end
end
